File: Python/Problem1.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices or len(prices) <= 1:
            return 0
        
        n = len(prices)
        
        first_profit = [0] * n
        min_price = prices[0]
        
        for i in range(1, n):
            min_price = min(min_price, prices[i])
            first_profit[i] = max(first_profit[i - 1], prices[i] - min_price)
        
        second_profit = [0] * n
        max_price = prices[-1]
        
        for i in range(n-2, -1, -1):
            max_price = max(max_price, prices[i])
            second_profit[i] = max(second_profit[i + 1], max_price - prices[i])
        
        max_profit = 0
        for i in range(n):
            max_profit = max(max_profit, first_profit[i] + second_profit[i])
            
        return max_profit
    
sol = Solution()
test1 = [3,3,5,0,0,3,1,4]
test2 = [1,2,3,4,5]
test3 = [7,6,4,3,1]
test4 = [2, 4, 1]

print(sol.maxProfit(test1))
print(sol.maxProfit(test2))
print(sol.maxProfit(test3))
print(sol.maxProfit(test4))

File: Java/Problem1.java
class Solution {
    public int maxProfit(int[] prices) {
        if(prices == null || prices.length < 2){
            return 0;
        } else {
            int size = prices.length, minPrice = prices[0], maxPrice = prices[size-1], maxTotalProfit = 0;
            int[] maxProfit1 = new int[size];
            int[] maxProfit2 = new int[size];

            for(int i=0; i<size; i++){
                minPrice = Math.min(minPrice, prices[i]);
                if(i == 0)
                    maxProfit1[i] = Math.max(maxProfit1[i], prices[i] - minPrice);
                else
                    maxProfit1[i] = Math.max(maxProfit1[i-1], prices[i] - minPrice);
            }

            for(int i=size-2; i >= 0; i--){
                maxPrice = Math.max(maxPrice, prices[i]);
                maxProfit2[i] = Math.max(maxProfit2[i+1], maxPrice - prices[i]);
            }

            for(int i=0; i<size; i++){
                maxTotalProfit = Math.max(maxTotalProfit, maxProfit1[i] + maxProfit2[i]);
            }
            
            return maxTotalProfit;
        }
    }
}